core-rust/state-manager/src/state_manager.rs

```diff
@@ -909,23 +910,21 @@ where
             .get_last_epoch_proof()
             .map(|epoch_proof| EpochTransactionIdentifiers::from(epoch_proof.ledger_header))
             .unwrap_or_else(EpochTransactionIdentifiers::pre_genesis);
-        let epoch_transactions_count = usize::try_from(
-            base_transaction_identifiers.state_version - epoch_identifiers.state_version,
-        )
-        .unwrap();
-
-        if base_transaction_identifiers.state_version != commit_request_start_state_version {
+        let base_state_version = base_transaction_identifiers.state_version;
+        if base_state_version != commit_request_start_state_version {
             panic!(
                 "Mismatched state versions - the commit request claims {} but the database thinks we're at {}",
-                commit_request_start_state_version, base_transaction_identifiers.state_version
+                commit_request_start_state_version, base_state_version
             );
         }
 
         let mut state_tracker = StateTracker::initial(base_transaction_identifiers);
         let mut committed_transaction_bundles = Vec::new();
         let mut substate_store_update = SubstateStoreUpdate::new();
         let mut state_tree_update = HashTreeUpdate::new();
-        let transaction_tree_len = epoch_transactions_count + 1; // starts with previous epoch root
```

Comment:
_Note from author:_

This was the off-by-one error, which did not take the "empty tree at start of epoch" case into account. Other places (that were also dealing with "per-epoch transaction trees") got this part right. So I extracted that logic into a common helper (kinda overkill, but that's the only way I could make it obvious; it is really hard to assert that thing in integration test).